sdlib/d/sync/sharedlock.d

```diff
@@ -0,0 +1,421 @@
+
+	lock.sharedLock();
+	assert(lock.count == 1);
+
+	// Attempt to exclusive lock while the shared lock is held.
+	import core.stdc.pthread;
+	import d.sync.atomic;
+	shared Atomic!uint state;
+
+	void* takeExclusiveLock() {
+		state.fetchAdd(1);
+		lock.exclusiveLock();
+		state.fetchAdd(1);
+		return cast(void*) cast(size_t) lock.count;
+	}
+
+	void* takeSharedLock() {
+		lock.sharedLock();
+		state.fetchAdd(1);
+		return null;
+	}
+
+	auto t1 = runThread(takeExclusiveLock);
+
+	{
+		bool triedExclusiveLock() {
+			return lock.count >= SharedLock.Exclusive;
+		}
+
+		lock.mutex.lock();
+		scope(exit) lock.mutex.unlock();
+
+		lock.mutex.waitFor(triedExclusiveLock);
+		assert(lock.count == SharedLock.Exclusive + 1);
+		assert(state.load() == 1);
+	}
+
+	auto t2 = runThread(takeSharedLock);
+
+	// Give t2 some time to try locking, sleep 0.5s
+	import core.stdc.unistd;
+	usleep(1000 * 500);
+	assert(lock.count == SharedLock.Exclusive + 1);
+	assert(state.load() == 1);
+
+	lock.sharedUnlock();
+	assert(lock.count == SharedLock.Exclusive);
+
+	void* ret;
+	pthread_join(t1, &ret);
+
+	assert(lock.mutex.isHeld());
+	assert(cast(size_t) ret == SharedLock.Exclusive);
+	assert(state.load() == 2);
+
+	lock.exclusiveUnlock();
+	pthread_join(t2, &ret);
+	assert(lock.count == 1);
+	assert(state.load() == 3);
+
+	lock.sharedUnlock();
+	assert(lock.count == 0);
+
+	void* run1() {
+		assert(state.load() == 0);
+
+		lock.exclusiveLock();
+		assert(lock.count == SharedLock.Exclusive, "Invalid lock state!");
+		assert(lock.mutex.isHeld(), "Mutex not held while exclusively locked!");
+		state.store(1);
+		lock.exclusiveUnlock();
+
+		while (state.load() != 2) {
+			import sys.posix.sched;
+			sched_yield();
+		}
+
+		return null;
+	}
+
+	void* run2() {
+		while (state.load() != 1) {
+			import sys.posix.sched;
+			sched_yield();
+		}
+
+		assert(state.load() == 1);
+
+		lock.exclusiveLock();
+		assert(lock.count == SharedLock.Exclusive, "Invalid lock state!");
+		assert(lock.mutex.isHeld(), "Mutex not held while exclusively locked!");
+		state.store(2);
+		lock.exclusiveUnlock();
+
+		return null;
+	}
+
+	state.store(0);
+	auto t3 = runThread(run1);
+	auto t4 = runThread(run2);
+
+	pthread_join(t3, &ret);
+	pthread_join(t4, &ret);
+
+	assert(!lock.mutex.isHeld(), "Mutex not held!");
+	assert(lock.count == 0, "Invalid lock state!");
+}
+
+unittest threadStressTest {
+	// Stress the lock by creating many threads that try both exclusive and
+	// shared locking.
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	import d.sync.atomic;
+	shared SharedLock lock;
+	shared Atomic!uint numSharedLocks;
+	shared Atomic!uint numExclusiveLocks;
+
+	enum MaxExclusive = 5;
+	enum MaxShared = 50;
```

Comment:
Why not much larger numbers?

Reply:
This was roughly 5 seconds runtime. I upped it to 20/200 and got to about 36 seconds runtime. High water mark went up.

If I went to 50/500, it took over a minute, but the high water mark was up to 49 locally.

So I left it at 20/200.